cmd/kubeadm/app/cmd/certs_test.go

```diff
@@ -482,3 +494,224 @@ kubernetesVersion: %s`,
+		kubeadmconstants.SuperAdminKubeConfigFileName,
+		kubeadmconstants.SchedulerKubeConfigFileName,
+		kubeadmconstants.ControllerManagerKubeConfigFileName,
+	}
+	for _, kubeConfig := range kubeConfigs {
+		if err := kubeconfigphase.CreateKubeConfigFile(kubeConfig, kdir, cfg); err != nil {
+			t.Fatalf("couldn't create kubeconfig %q: %v", kubeConfig, err)
+		}
+	}
+
+	// A minimal kubeadm config with just enough values to avoid triggering
+	// auto-detection of config values at runtime.
+	var kubeadmConfig = fmt.Sprintf(`
+apiVersion: %[1]s
+kind: ClusterConfiguration
+certificatesDir: %s
+kubernetesVersion: %s`,
+		kubeadmapiv1.SchemeGroupVersion.String(),
+		cfg.CertificatesDir,
+		kubeadmconstants.MinimumControlPlaneVersion.String())
+
+	// Write the minimal kubeadm config to a file
+	customConfigPath := kdir + "/kubeadm.conf"
+	f, err := os.Create(customConfigPath)
+	require.NoError(t, err)
+	_, err = f.Write([]byte(kubeadmConfig))
+	require.NoError(t, err)
+
+	// fakeClientSetFromFile returns a fake clientset with kubeadm config map
+	var fakeClientSetFromFile = func(_ string) (kubernetes.Interface, error) {
+		client := fakeclientset.NewSimpleClientset()
+		client.PrependReactor("get", "configmaps", func(action clientgotesting.Action) (bool, runtime.Object, error) {
+			getAction := action.(clientgotesting.GetAction)
+			if getAction.GetNamespace() == metav1.NamespaceSystem && getAction.GetName() == kubeadmconstants.KubeadmConfigConfigMap {
+				cm := &v1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      kubeadmconstants.KubeadmConfigConfigMap,
+						Namespace: metav1.NamespaceSystem,
+					},
+					Data: map[string]string{
+						kubeadmconstants.ClusterConfigurationConfigMapKey: dedent.Dedent(kubeadmConfig),
+					},
+				}
+				return true, cm, nil
+			}
+			// Return an empty config map for kubelet and kube-proxy components.
+			return true, &v1.ConfigMap{}, nil
+		})
+		return client, nil
+	}
+
+	// Select a certificate used to simulate a missing certificate file
+	brokenCertName := kubeadmconstants.APIServerCertAndKeyBaseName
+	brokenCertPath, _ := pkiutil.PathsForCertAndKey(cfg.CertificatesDir, brokenCertName)
+
+	type testCase struct {
+		name           string
+		config         string
+		output         string
+		brokenCertName string
+	}
+
+	var runTestCase = func(t *testing.T, tc testCase) {
+		var output bytes.Buffer
+		cmd := newCmdCertsExpiration(&output, kdir)
+		args := []string{
+			fmt.Sprintf("--cert-dir=%s", cfg.CertificatesDir),
+		}
+		if tc.config != "" {
+			args = append(args, fmt.Sprintf("--config=%s", tc.config))
+		} else {
+			clientSetFromFile = fakeClientSetFromFile
+			defer func() {
+				clientSetFromFile = kubeconfigutil.ClientSetFromFile
+			}()
+		}
+		if tc.output != "" {
+			args = append(args, fmt.Sprintf("-o=%s", tc.output))
+		}
+		cmd.SetArgs(args)
+		require.NoError(t, cmd.Execute())
+
+		switch tc.output {
+		case "json":
+			assert.True(t, yamlutil.IsJSONBuffer(output.Bytes()))
```

Comment:
i think we can remove this check and the yamlutil import too.
json.Unmarshal should catch if the output is not JSON, correct?